### ISSUE TITLE
Sketchfab auth broken

### ISSUE THREAD

**Issue by wekitecs** [WEKIT-ECS/MIRAGE-XR#590]
In GitLab by @wild on Sep 9, 2021, 18:01

The sketchfab auth is broken - it opens the browser window, but this seems to kill the app in the background, starting a new app upon return. But even the fresh start does not pick up the auth token, and the app remains unauthorised.

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 16, 2021, 14:25

duplicate of #566

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 16, 2021, 14:26

hard to resolve, because of #565

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 24, 2021, 12:18

Educated guess: 

The Microsoft documentation of the HttpClient class says: "HttpClient is intended to be instantiated once and re-used throughout the life of an application. Instantiating an HttpClient class for every request will exhaust the number of sockets available under heavy loads. This will result in SocketException errors." https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-5.0

Therefore, the script MirageXR/Common/Scripts/Sketchfab/Sketchfab.cs should have a 
```
    private static readonly HttpClient HttpClient;
```
at the beginning, and then loose the "var " in line 368.

This should funnily resolve it, as the problem might actually come from disposing the HttpClient because of using the "using(){}" structure in line 363. Using "using(...){}" disposes of the vars declared inside of the block, which is effectively when the async call is dispatched. So, the browser call is correctly dispatched, but its response cannot return, as the client does no longer exist.

See here:
https://www.aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 24, 2021, 12:26

If this does not solve it, here are a few more ideas:

It is possible to force HttpClient to use a specific implementation. By default, it seems to use HttpWebRequest, but - and the documentation mentions this explicitly - for authentication WebRequestHandler can be used to configure settings.

> "By default on .NET Framework and Mono, HttpWebRequest is used to send requests to the server. This behavior can be modified by specifying a different channel in one of the constructor overloads taking a HttpMessageHandler instance as parameter. If features like authentication or caching are required, WebRequestHandler can be used to configure settings and the instance can be passed to the constructor. The returned handler can be passed to one of the constructor overloads taking a HttpMessageHandler parameter."

This property of HttpWebRequest might be relevant: https://docs.microsoft.com/en-us/dotnet/api/system.net.httpwebrequest.keepalive?view=net-5.0

Mind it, this would only be hololens platform - as other implementations will be used for Android and iOS.

---

If this all fails, it may be worth switching to the Unity class equivalent - instead of the .net implementation HttpClient - using UnityWebRequest instead, which uses a different invocation method (see (https://docs.unity3d.com/Manual/UnityWebRequest-SendingForm.html)[here]): 

```
using UnityEngine;
using UnityEngine.Networking;
using System.Collections;
using System.Collections.Generic;

public class MyBehavior : MonoBehaviour
{
    void Start()
    {
        StartCoroutine(Upload());
    }

    IEnumerator Upload()
    {
        List<IMultipartFormSection> formData = new List<IMultipartFormSection>();
        formData.Add(new MultipartFormDataSection("field1=foo&field2=bar"));
        formData.Add(new MultipartFormFileSection("my file data", "myfile.txt"));

        UnityWebRequest www = UnityWebRequest.Post("https://www.my-server.com/myform", formData);
        yield return www.SendWebRequest();

        if (www.result != UnityWebRequest.Result.Success)
        {
            Debug.Log(www.error);
        }
        else
        {
            Debug.Log("Form upload complete!");
        }
    }
}
```

But Alex has introduced it for the MoodleManager (and others), and I am sure he has had reasons for that (it looks like it is very cross-platform and very flexible).

For debugging: This here mentions that you can try-catch the specific request cancellation event, may be useful:
https://stackoverflow.com/questions/29179848/httpclient-a-task-was-cancelled (first highlighted answer 319)

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 24, 2021, 12:38

which object uses MirageXR/Common/Scripts/Sketchfab/Sketchfab.cs? is this in the mobile player, because I don't see it used by the 'ModelSelector' prefab in the Player scene.

in my branch (forked from develop yesterday), i see no dependencies for this script:

![image](https://platform.xr4all.eu/wekit-ecs/mirage-xr/uploads/50180cb255abcbbe3c1b29ca323e1f6b/image.png)

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 24, 2021, 14:08

right, i have narrowed the start of the error down to one line - 114, in SketchfabManager.cs:

`service.OpenLoginPage();`

showing the opening of the page is connected to the closure of the app. It might go as low as the windows process level, but hard to say, as I didn't get that deep into the i5 toolkit's functionality.

@BenediktHensen perhaps you could take a look to see if there have been changes to the Windows Holographic OS that might be in conflict with the implementation in the i5 toolkit?

i found this, but not sure if it's relevant: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1468. I can continue to look for other routes but, frankly, am a little stuck on this.

p.s. I tried a global declaration of the OpenIDConnectService, to no avail.

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 27, 2021, 11:10

as suggested by BH, this could also be related to the RAM usage of the application. if this had gone over a threshold, it might close the app rather than maintain it while the browser is opened. i am investigating this now

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 27, 2021, 11:28

this might have something to do with things...

![image](https://platform.xr4all.eu/wekit-ecs/mirage-xr/uploads/95392a3e424c90ccb2022194b69df278/image.png)

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 27, 2021, 11:51

a quick check in Unity profiler:

![image](https://platform.xr4all.eu/wekit-ecs/mirage-xr/uploads/a7bd542b7cb80dffd8e32d322a5ff812/image.png)

shows 0.98 GB total memory use. This is at application start, where the activity list is showing, and so likely to go up as additional functions are used

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 27, 2021, 12:28

the device portal also gives some indication of memory use and performance tracing. This snip is taken just before the 'authorise sketchfab' button is pressed:

![image](https://platform.xr4all.eu/wekit-ecs/mirage-xr/uploads/c9c33db92fd6cfd960f2f1e559b56c46/image.png)

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 27, 2021, 12:47

Can you try out in a breakout project? This could subsequently be used to have a test embedded that can track this?

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 27, 2021, 12:58

Here are some more tips from Benedikt: https://codereality.net/ar-for-eu-book/chapter/development/tools/unity/advanced/performanceProfiling/

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 27, 2021, 12:58

I am not sure I follow. i am using the profiler to check memory usage, but it's not something that is in the code project. I see a couple of issues using unit tests to assertain performance:

1) Since the memory usage will fluctuate throughout the use of the app, taking a snapshot would not really guarantee that other functions work.

2) The unit tests also run in their own dummy scene and so avoid loading a lot of the functions that would use memory, so the value would likely not be representative of the actual usage.

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 27, 2021, 13:11

the breakout project is set up and i have moved #592 to it

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 27, 2021, 14:29

Misunderstanding: with 'test' I meant: write a test to check if sketchfab works - which is the same thing you need for a breakout project anyways. This way, when the test is executed, it should be able to catch such problems? Instead of forking to web-browser, it may be best to use a batch script that can be tested (and where the return value is clear)...

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 27, 2021, 19:36

i have packaged the sketchfab parts and tested them under less load - profiler shows 680-700 MB (of 2048 MB total) - and I can see both the webpage and the authorise screen simulataneously, though only for a few seconds, after which the app crashes again.

![image](https://platform.xr4all.eu/wekit-ecs/mirage-xr/uploads/be94610316fd2257d37f6ecd8b9f9aa5/image.png)

Looking at the device portal at the moment just before the crash shows the following, and there is no sudden spike that corresponds with the app closing. The column 'commit size' matches the value shown by the MRTK profiler.

*Mirage=Blue, MS Edge=Orange*

![image](https://platform.xr4all.eu/wekit-ecs/mirage-xr/uploads/9a206a48b1d0d2f899846e9cf48a68dc/image.png)

I am not so concerned that there are multiple entries for each app, though it seems odd that there are four for 'msedge' each using 206.6 MB. (When i load Edge on Windows Desktop, this is not the case).

In conclusion, it doesn't seem to me that a lack of RAM is causing the problem. Though one possible limit might be 900 MB, the number is not getting very close and (even if multiple browsers are loading) the device still has some spare.

That said, if you add together the entries for either Mirage or Edge, they total more than 900 MB - 986 and 972, respectively - which could be problematic or could be the reason that there are multiples items.

**Comment by wekitecs** [comment]
In GitLab by @BenediktHensen on Sep 28, 2021, 19:08

I am checking this as well. Currently trying a roll-back to Unity 2019 to see if it is related to the newer Unity version and I am creating a separate experimentation app to go over each step of the login.

**Comment by wekitecs** [comment]
In GitLab by @BenediktHensen on Sep 29, 2021, 22:31

I created the desktop project to test the login. It is definitely a crash caused in the Unity.dll that happens as soon as the browser is opened.

**Comment by wekitecs** [comment]
In GitLab by @BenediktHensen on Sep 29, 2021, 22:43

This is actually a known problem and Unity bug: https://forum.unity.com/threads/application-openurl-url-causes-crash-on-uwp.1024660/

**Comment by wekitecs** [comment]
In GitLab by @BenediktHensen on Sep 29, 2021, 23:02

The problem occurs as soon as `Application.OpenURL` is called and it throws the following exception:

```
Exception thrown at 0x00007FFDFC77E197 (UnityPlayer.dll) in miragexr.exe: 0xC0000005: Access violation reading location 0x0000000000000000.
```

The next step would be to try and update to a newer Unity version to check if they already fixed the error. If it is still there, we need to find an alternative to Application.OpenURL.

**Comment by wekitecs** [comment]
In GitLab by @BenediktHensen on Sep 30, 2021, 10:29

The bug still persists on Unity 2020.3.19f1. I will check Unity's issue tracker to see if they are aware of the bug.

**Comment by wekitecs** [comment]
In GitLab by @BenediktHensen on Sep 30, 2021, 11:03

I filed a bug report at Unity.

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 30, 2021, 12:06

Ok, thanks. Hopefully they know what this is. Crazy thought: recently, the eye tracker was activated in the project. Could it be that there is a buggy eye-tracker routine? That's one thing that has changed for sure, its far fetched, but ... maybe...?

**Comment by wekitecs** [comment]
In GitLab by @BenediktHensen on Sep 30, 2021, 12:09

Unfortunately no. I also created a HoloLens-independent test project to submit the bug report. The error already happens in UWP desktop builds without any HoloLens functionality.

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 30, 2021, 12:18

In which case, I would almost propose to release 1.7 - and list this as a known bug. The workaround for this is: install 1.6, login on sketchfab, install 1.7, and your are still logged in. HOWEVER: currently, the preview builds cannot download models. The preview works, and the pop up appears that asks whether I want to download, but the download fails. This would need to be fixed first.

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 30, 2021, 12:29

Actually: in current preview (1.6.124) not even the thumbnails work on hololens. the popup dialogue asking whether you want to download appears, but nothing after.

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 30, 2021, 13:32

i am just checking the branch in !323 which should resolve this

**Comment by wekitecs** [comment]
In GitLab by @william.guest on Sep 30, 2021, 14:06

the thumbnail fix is ready to be merged (!323)

**Comment by wekitecs** [comment]
In GitLab by @wild on Sep 30, 2021, 16:29

unassigned @william.guest
